source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -167,29 +167,21 @@ OptRef<const LocalRateLimiterImpl::LocalDescriptorImpl> LocalRateLimiterImpl::de
 
 bool LocalRateLimiterImpl::requestAllowed(
     absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
-
-  bool allow = requestAllowedHelper(tokens_);
-  // Global token is not enough. Since global token is not sorted, so we suggest it should be
-  // larger than other descriptors.
-  if (!allow) {
-    return allow;
-  }
-
   if (!descriptors_.empty() && !request_descriptors.empty()) {
     for (const auto& descriptor : sorted_descriptors_) {
       for (const auto& request_descriptor : request_descriptors) {
         if (descriptor == request_descriptor) {
-          allow &= requestAllowedHelper(*descriptor.token_state_);
           // Descriptor token is not enough.
-          if (!allow) {
-            return allow;
+          if (!requestAllowedHelper(*descriptor.token_state_)) {
+            return false;
           }
           break;
         }
       }
     }
   }
-  return allow;
+  // Since global token is not sorted, so we suggest it should be larger than other descriptors.
+  return requestAllowedHelper(tokens_);
```

Comment:
Typically, there is no perfect approach for this problem. Seems this a little better than the old behavior.

But I am not sure if a runtime guard is necessary for this behavior change?  cc @mattklein123 

cc @StarryVae for you are working on local ratelimit related things.

Reply:
I'm on the fence about a guard. It's probably OK without. I don't understand the comment though about sorting and size. Can you expand the comment?

Reply:
Sure. I've added an extra comment to give more context about how descriptor tokens are consumed.

```c++
bool LocalRateLimiterImpl::requestAllowed(                                                
    absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {             
  // Matched descriptors will be sorted by tokens per second and tokens consumed in order.
  // In most cases, if one of them is limited the remaining descriptors will not consume  
  // their tokens.                                                                        
  if (!descriptors_.empty() && !request_descriptors.empty()) {                            
    for (const auto& descriptor : sorted_descriptors_) {                                  
      for (const auto& request_descriptor : request_descriptors) {                        
        if (descriptor == request_descriptor) {                                           
          // Descriptor token is not enough.                                              
          if (!requestAllowedHelper(*descriptor.token_state_)) {                          
            return false;                                                                 
          }                                                                               
          break;                                                                          
        }                                                                                 
      }                                                                                   
    }                                                                                     
  }                                                                                       
  // Since global tokens are not sorted, it should be larger than other descriptors.      
  return requestAllowedHelper(tokens_);                                                   
}                                                                                         
```